### PR TITLE
[FIX] base: PNGs are converted to a WEB palette when downsized

### PR DESCRIPTION
**Pull request by agau-odoo**
When uploading a png image ir_attachment, the image is not modified if its resolution is under the maximum 1920x1920. However, if the resolution is bigger, it both gets resized and is converted to a WEB palette, which visibly degrades the quality of the image.

A workaround for this is to resize the image locally to be max 1920 on either dimensions and then upload it, which effectively bypasses this special treatment.

Steps to reproduce:
- Go to the website app
- Add a Text - Image snippet
- Double click the image
- Upload a .png image of a resolution strictly greater than 1920 in either width or height

Old behavior: the png is visibly degraded
New behavior: the png is not visibly degraded

opw-3935533

### CODE REVIEW

**Review by rco-odoo — COMMENTED** [review]
The code makes sense to me, but I don't consider myself competent for reviewing this change.

File: odoo/addons/base/models/ir_attachment.py [hunk 1]
```diff
@@ -336,7 +336,10 @@ def _postprocess_contents(self, values):
                     nw, nh = map(int, max_resolution.split('x'))
                     if w > nw or h > nh:
                         img = img.resize(nw, nh)
-                        quality = int(ICP('base.image_autoresize_quality', 80))
+                        if _subtype == 'png':
+                            quality = 0  # Do not affect PNGs color palette
+                        else:
+                            quality = int(ICP('base.image_autoresize_quality', 80))
```
Comment (seb-odoo) [inline odoo/addons/base/models/ir_attachment.py:342]: This will make the `quality` param only used for `jpeg`, so we might want to make that explicit to avoid the guessing game, rather than discarding `png`? Especially since the value of 80 looks tailored to `jpeg` specifically.

**Review by seb-odoo — COMMENTED** [review]
Hello,

This is not really a fix, but a functional change to prioritize quality over file size for PNG.
Not implying I'm in favor or against the change, but we have to keep it in mind.

The "bad" quality might help users realize they need to use a better format, while still preventing excessive file size when they don't.

Alternative solution for users would be to provide a small enough image, use `jpeg` or even better `webp`.

There are many more alternatives mentioned on the original commit: https://github.com/odoo/odoo/commit/d9ce0507960f247e1187baf7bd8399f90be237aa

I agree the inconsistency of quality change based on whether there is a resize happening or not is confusing, and even for jpeg actually. We could also consider to always change quality, even when there is no resize.

But I'm pretty sure all of that was known when making the original task and commit, and the decision was made to do it like it is currently for good reasons. So I'm not sure there is a good enough reason to change that after 3 years, on an almost no longer supported version of Odoo.

But I have no strong opinion either way.

cc @JKE-be

**Comment by agau-odoo** [comment]
> The "bad" quality might help users realize they need to use a better format, while still preventing excessive file size when they don't.
I agree that we might want to push users towards using more appropriate image formats, however I believe that a clear warning/popup message would be much more user-friendly.

> I agree the inconsistency of quality change based on whether there is a resize happening or not is confusing, and even for jpeg actually. We could also consider to always change quality, even when there is no resize.
I do like this solution too, however this also needs to come with some kind of UX warning for the user too.

File: odoo/addons/base/models/ir_attachment.py [hunk 2]
```diff
@@ -336,7 +336,10 @@ def _postprocess_contents(self, values):
                     nw, nh = map(int, max_resolution.split('x'))
                     if w > nw or h > nh:
                         img = img.resize(nw, nh)
-                        quality = int(ICP('base.image_autoresize_quality', 80))
+                        if _subtype == 'png':
```
Comment (qsm-odoo) [inline odoo/addons/base/models/ir_attachment.py:339]: > supported_subtype = ICP('base.image_autoresize_extensions', 'png,jpeg,bmp,tiff').split(',')

I see that webp is not listed in there in master, is that another bug or am I missing something?

File: odoo/addons/base/models/ir_attachment.py (hunk 1 above)
Comment (qsm-odoo) [reply odoo/addons/base/models/ir_attachment.py:342]: 1. If it does not impact the others (bmp, etc), should not the code say `!== jpeg` @agau-odoo ?
2. Not sure it's really that needed to make it more explicit... are not those ICP you can only know about looking at this code? But I guess we can rename it in master if we merge this... but I feel like we only risk annoying the 3 people that modified this in their DB, for nothing, so I would not do it :shrug: I am not a format expert but I don't think any dev would be unhappy that "quality" only applies for jpeg given our supported formats (not sure about webp though)?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
> The "bad" quality might help users realize they need to use a better format, while still preventing excessive file size when they don't.

Good point but if I were to guess, I would say that more people will see this as a bug rather than a hint to use another format. Certainly most of our users using the website builder (and even more using Odoo in general) has no idea of the difference between jpeg and png.

And for the legitimate cases of using a png that big, feels right to not optimize it.

> I agree the inconsistency of quality change based on whether there is a resize happening or not is confusing, and even for jpeg actually. We could also consider to always change quality, even when there is no resize.

Meh, maybe. Or even no quality change at all? Not sure how it is applied, I guess we need to force a quality? Not sure it is worth the debate, I'd not touch what we don't need to (as you said, there were debates years ago, I am sure we don't need reviving them).

> So I'm not sure there is a good enough reason to change that after 3 years, on an almost no longer supported version of Odoo.

Well, it is a reported bug fix, and as always we consider fixing it where we can. But yes, we could only target 17.0... I'd say we don't have to care that much and 15.0 is fine though, but not important.
In the end, we are talking about someone uploading a heavily colored PNG with a > 1920 size... I don't think we really risk much changing the code for that usecase.

Let's go with this unless someone disagrees:
- The customer is in >17.0, for once, let's re-target this PR to 17.0
- Basically, keep this diff (see comments below though, maybe worth changing the condition?)
- Create a task in master to consider adding a warning suggesting better image format when using a large PNG

**Review by qsm-odoo — APPROVED** [review]
This task was totally forgotten... I'd say let's go with it as it is. I tested it, it seems fine and I judge the risk/reward ratio is good.

@robodoo r+
